File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
# 1. 구할 수 있는 영문쌍만 필터링 & 대문자 치환
# 2. set를 이용해 교집합 & 합집합 구하기
# 3. min, max 구해서 자카드 유사도 구하기
import re
import math
def solution(str1, str2):
    
    str1_set = [str1[i:i+2].upper() for i in range(len(str1)-1) if not re.findall('[^a-zA-Z]', str1[i:i+2])]
    str2_set = [str2[i:i+2].upper() for i in range(len(str2)-1) if not re.findall('[^a-zA-Z]', str2[i:i+2])]
    
    intersection = set(str1_set) & set(str2_set)
    union = set(str1_set) | set(str2_set)
    
    # 공집합인 경우
    if not union:
        return 65536
    
    jac_intersection = [min(str1_set.count(i), str2_set.count(i)) for i in intersection]
    jac_union = [max(str1_set.count(u), str2_set.count(u)) for u in union]
    
    return math.floor(sum(jac_intersection) / sum(jac_union) * 65536)
        

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
function getSet(str) {
    const answer = []
    const upper = str.toUpperCase()
    
    for (let i = 0; i < upper.length - 1; i++) {
        if (('A' <= upper[i] && upper[i] <= 'Z') && ('A' <= upper[i + 1] && upper[i + 1] <= 'Z')) {
            answer.push(upper.substr(i, 2))
        }
    }
    
    return answer
}

function solution(str1, str2) {
    const multi_set1 = getSet(str1)
    const multi_set2 = getSet(str2)
    const set1 = new Set(multi_set1)
    const set2 = new Set(multi_set2)
    
    // 공집합인 경우
    if (set1.size === 0 && set2.size === 0) {
        return (1 * 65536)
    }
    
    // 합집합 / 교집합 구하기
    const union = new Set()
    const intersection = new Set()
    
    set1.forEach((elem) => {
        union.add(elem)
        if (set2.has(elem)) {
            intersection.add(elem)
        }
    })
    
    set2.forEach((elem) => {
        union.add(elem)
    })
    
    console.log(union, intersection)
    
    // 각 multi set에 원소 개수 구하기
    // 1. 합집합
    let union_cnt = 0
    for (const u of union) {
        const multi_set1_cnt = multi_set1.reduce((cnt, elem) => cnt + (elem === u), 0)
        const multi_set2_cnt = multi_set2.reduce((cnt, elem) => cnt + (elem === u), 0)
        
        union_cnt = union_cnt + Math.max(multi_set1_cnt, multi_set2_cnt)
    }
    
    // 2. 교집합
    let intersection_cnt = 0
    for (const i of intersection) {
        const multi_set1_cnt = multi_set1.reduce((cnt, elem) => cnt + (elem === i), 0)
        const multi_set2_cnt = multi_set2.reduce((cnt, elem) => cnt + (elem === i), 0)
        
        intersection_cnt = intersection_cnt + Math.min(multi_set1_cnt, multi_set2_cnt)
    }
    
    return Math.floor((intersection_cnt / union_cnt) * 65536)
}